File: composeApp/src/androidMain/kotlin/com/dnfapps/arrmatey/ui/screens/EditInstanceScreen.kt
package com.dnfapps.arrmatey.ui.screens

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.WindowInsets
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.statusBars
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.ArrowBack
import androidx.compose.material.icons.filled.Delete
import androidx.compose.material3.AlertDialog
import androidx.compose.material3.Button
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.IconButtonDefaults
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.material3.TopAppBar
import androidx.compose.material3.TopAppBarDefaults
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.input.nestedscroll.nestedScroll
import androidx.compose.ui.unit.dp
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import com.dnfapps.arrmatey.arr.viewmodel.EditInstanceViewModel
import com.dnfapps.arrmatey.database.dao.InsertResult
import com.dnfapps.arrmatey.navigation.SettingsNavigation
import com.dnfapps.arrmatey.shared.MR
import com.dnfapps.arrmatey.utils.koinInjectParams
import com.dnfapps.arrmatey.utils.mokoString
import com.dnfapps.arrmatey.utils.navigationBarBottomInset
import kotlinx.coroutines.launch
import org.koin.compose.koinInject

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun EditInstanceScreen(
    id: Long,
    viewModel: EditInstanceViewModel = koinInjectParams(id),
    settingsNav: SettingsNavigation = koinInject<SettingsNavigation>()
) {
    val scope = rememberCoroutineScope()
    val scrollBehavior = TopAppBarDefaults.pinnedScrollBehavior()

    val uiState by viewModel.uiState.collectAsStateWithLifecycle()
    val instance by viewModel.instance.collectAsStateWithLifecycle()

    var confirmDelete by remember { mutableStateOf(false) }
    var saveClicked by remember { mutableStateOf(false) }

    LaunchedEffect(uiState.testResult) {
        if (uiState.testResult == true && saveClicked) {
            viewModel.updateInstance()
        }
    }

    LaunchedEffect(uiState.editResult) {
        if (uiState.editResult is InsertResult.Success) {
            settingsNav.popBackStack()
        }
    }

    Scaffold(
        modifier = Modifier.nestedScroll(scrollBehavior.nestedScrollConnection),
        topBar = {
            TopAppBar(
                scrollBehavior = scrollBehavior,
                title = { Text(text = mokoString(MR.strings.edit_instance)) },
                navigationIcon = {
                    IconButton(
                        onClick = { settingsNav.popBackStack() }
                    ) {
                        Icon(
                            imageVector = Icons.AutoMirrored.Default.ArrowBack,
                            contentDescription = mokoString(MR.strings.back)
                        )
                    }
                },
                actions = {
                    IconButton(
                        onClick = {
                            confirmDelete = true
                        },
                        colors = IconButtonDefaults.iconButtonColors(
                            containerColor = MaterialTheme.colorScheme.errorContainer,
                            contentColor = MaterialTheme.colorScheme.onErrorContainer
                        ),
                        modifier = Modifier.padding(end = 4.dp)
                    ) {
                        Icon(
                            imageVector = Icons.Default.Delete,
                            contentDescription = null
                        )
                    }

                    Button(
                        onClick = {
                            scope.launch {
                                saveClicked = true
                                viewModel.testConnection()
                            }
                        },
                        modifier = Modifier.padding(end = 12.dp)
                    ) {
                        Text(text = mokoString(MR.strings.save))
                    }
                }
            )
        },
        contentWindowInsets = WindowInsets.statusBars
    ) { contentPadding ->
        Column(
            modifier = Modifier
                .padding(contentPadding)
                .fillMaxSize()
                .verticalScroll(rememberScrollState())
                .padding(horizontal = 16.dp)
                .padding(bottom = navigationBarBottomInset() + 16.dp),
            verticalArrangement = Arrangement.spacedBy(16.dp)
        ) {
            instance?.let { instance ->
                ArrConfigurationScreen(
                    instanceType = instance.type,
                    uiState = uiState,
                    onApiEndpointChanged = { viewModel.setApiEndpoint(it) },
                    onApiKeyChanged = { viewModel.setApiKey(it) },
                    onInstanceLabelChanged = { viewModel.setInstanceLabel(it) },
                    onIsSlowInstanceChanged = { viewModel.setIsSlowInstance(it) },
                    onCustomTimeoutChanged = { viewModel.setCustomTimeout(it) },
                    onHeadersChanged = { viewModel.updateHeaders(it) },
                    onTestConnection = { viewModel.testConnection() },
                    onLocalNetworkEnabledChanged = { viewModel.setLocalNetworkEnabled(it) },
                    onLocalNetworkUrlChanged = { viewModel.setLocalNetworkUrl(it) },
                    onLocalNetworkSsidChanged = { viewModel.setLocalNetworkSsids(it) },
                    onTestLocalConnection = { viewModel.testLocalConnection() },
                    onToggleNotificationsEnabled = { viewModel.toggleNotificationsEnabled() }
                )
            }

            if (confirmDelete) {
                AlertDialog(
                    onDismissRequest = { confirmDelete = false},
                    confirmButton = {
                        TextButton(
                            onClick = {
                                scope.launch {
                                    instance?.let { instance ->
                                        viewModel.deleteInstance(instance)
                                        settingsNav.popBackStack()
                                    }
                                }
                            }
                        ) { Text(mokoString(MR.strings.yes)) }
                    },
                    dismissButton = {
                        TextButton(
                            onClick = {
                                confirmDelete = false

                            }
                        ) { Text(mokoString(MR.strings.no)) }
                    },
                    title = { Text(mokoString(MR.strings.confirm)) },
                    text = {
                        Text(mokoString(MR.strings.confirm_delete_instance, instance?.label ?: ""))
                    }
                )
            }
        }
    }
}

File: iosApp/iosApp/Views/Screens/EditInstanceScreen.swift
//
//  EditInstanceScreen.swift
//  iosApp
//
//  Created by Owen LeJeune on 2025-12-17.
//

import Shared
import SwiftUI

struct EditInstanceScreen: View {
    private let id: Int64
    
    @Environment(\.dismiss) var dismiss
    
    @ObservedObject private var viewModel: EditInstanceViewModelS
    
    @State private var showDeleteConfirmation: Bool = false
    @State private var saveClicked: Bool = false
    private let onSaveSuccess: () -> Void
    
    init(id: Int64, onSaveSuccess: @escaping () -> Void) {
        self.id = id
        self.viewModel = EditInstanceViewModelS(id)
        self.onSaveSuccess = onSaveSuccess
    }
    
    private var uiState: AddInstanceUiState {
        viewModel.uiState
    }
    
    private var instance: Instance? {
        viewModel.instance
    }
    
    private var testResult: Bool? {
        uiState.testResult?.boolValue
    }
    
    private var editSuccess: Bool {
        uiState.editResult is InsertResultSuccess
    }
    
    
    var body: some View {
        content
            .onChange(of: viewModel.uiState.testResult) { _, newValue in
                if newValue?.boolValue == true && saveClicked {
                    viewModel.updateInstance()
                }
            }
            .onChange(of: editSuccess) { _, newValue in
                if newValue {
                    dismiss()
                    onSaveSuccess()
                }
            }
    }
    
    @ViewBuilder
    private var content: some View {
        ArrConfigurationView(
            uiState: uiState,
            onApiEndpointChanged: { viewModel.setApiEndpoint($0) },
            onApiKeyChanged: { viewModel.setApiKey($0) },
            onBasicAuthChanged: { viewModel.setBasicAuthEnabled($0) },
            onInstanceLabelChanged: { viewModel.setInstanceLabel($0) },
            onIsSlowInstanceChanged: { viewModel.setIsSlowInstance($0) },
            onCustomTimeoutChanged: { viewModel.setCustomTimeout($0) },
            onHeadersChanged: { viewModel.updateHeaders($0) },
            onTestConnection: { viewModel.testConnection() },
            onLocalNetworkEnabledChanged: { viewModel.setLocalNetworkEnabled($0) },
            onLocalNetworkUrlChanged: { viewModel.setLocalNetworkUrl($0) },
            onLocalNetworkSsidsChanged: { viewModel.setLocalNetworkSsids($0) },
            onTestLocalConnection: { viewModel.testLocalConnection() },
            onToggleNotificationsEnabled: { viewModel.toggleNotificationsEnabled() },
            onDismissInfoCard: { _ in },
            showInfoCard: false,
            showInstancePicker: false,
            instanceType: .constant(instance?.type ?? .sonarr),
            showError: $viewModel.showError
        )
        .toolbar { toolbarContent }
        .alert(MR.strings().confirm.localized(), isPresented: $showDeleteConfirmation, presenting: viewModel.instance) { instance in
            confirmDeleteButtons(instance)
        } message: { instance in
            Text(MR.strings().confirm_delete_instance.formatted(args: [instance.label]))
        }
    }
    
    @ViewBuilder
    private func confirmDeleteButtons(_ instance: Instance) -> some View {
        Button(MR.strings().yes.localized(), role: .destructive) {
            viewModel.delete(instance)
            dismiss()
        }
        Button(MR.strings().no.localized(), role: .cancel) {
            showDeleteConfirmation = false
        }
    }
    
    @ToolbarContentBuilder
    private var toolbarContent: some ToolbarContent {
        if instance != nil {
            ToolbarItem(placement: .primaryAction) {
                Image(systemName: "trash")
                    .imageScale(.medium)
                    .foregroundColor(.red)
                    .onTapGesture {
                        showDeleteConfirmation = true
                    }
            }
            ToolbarItem(placement: .primaryAction) {
                Button(MR.strings().save.localized()) {
                    Task {
                        saveClicked = true
                        viewModel.testConnection()
                    }
                }
            }
        }
    }

}
